File: ios/PolyfenceModule.swift
import React
import UIKit
import CoreLocation
import PolyfenceCore

@objc(PolyfenceModule)
class PolyfenceModule: RCTEventEmitter, PolyfenceCoreDelegate {

    private var locationTracker: LocationTracker?
    private var zonePersistence: ZonePersistence?
    private var hasListeners = false

    override func supportedEvents() -> [String] {
        return ["onLocation", "onGeofenceEvent", "onError", "onPerformance"]
    }

    override static func requiresMainQueueSetup() -> Bool {
        return false
    }

    override func startObserving() {
        hasListeners = true
    }

    override func stopObserving() {
        hasListeners = false
    }

    @objc(initialize:resolver:rejecter:)
    func initialize(config: NSDictionary?, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            if let configDict = config?["config"] as? [String: Any],
               let version = configDict["pluginVersion"] as? String {
                PolyfenceDebugCollector.shared.setPluginVersion(version)
            }

            zonePersistence = ZonePersistence()
            locationTracker = LocationTracker()
            locationTracker?.setDelegate(self)

            LocationTracker.shared.setBridgePlatform("react-native")

            if let configDict = config?["config"] as? [String: Any],
               let disableAlerts = configDict["disableAlertNotifications"] as? Bool {
                locationTracker?.setAlertNotificationsEnabled(!disableAlerts)
            }

            resolve(nil)
        } catch {
            NSLog("PolyfenceModule: Initialize failed: %@", error.localizedDescription)
            reject("INITIALIZATION_FAILED", error.localizedDescription, error)
        }
    }

    @objc(startTracking:rejecter:)
    func startTracking(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let tracker = locationTracker else {
                throw NSError(domain: "PolyfenceModule", code: 1, userInfo: [NSLocalizedDescriptionKey: "Location tracker not initialized"])
            }
            tracker.startTracking()
            sendStatus(trackingEnabled: true)
            resolve(nil)
        } catch {
            NSLog("PolyfenceModule: Start tracking failed: %@", error.localizedDescription)
            reject("START_TRACKING_FAILED", error.localizedDescription, error)
        }
    }

    @objc(stopTracking:rejecter:)
    func stopTracking(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let tracker = locationTracker else {
                throw NSError(domain: "PolyfenceModule", code: 2, userInfo: [NSLocalizedDescriptionKey: "Location tracker not initialized"])
            }
            tracker.stopTracking()
            sendStatus(trackingEnabled: false)
            resolve(nil)
        } catch {
            NSLog("PolyfenceModule: Stop tracking failed: %@", error.localizedDescription)
            reject("STOP_TRACKING_FAILED", error.localizedDescription, error)
        }
    }

    @objc(addZone:resolver:rejecter:)
    func addZone(zoneData: NSDictionary?, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let zoneDict = zoneData as? [String: Any],
                  let zoneId = zoneDict["id"] as? String,
                  let zoneName = zoneDict["name"] as? String else {
                throw NSError(domain: "PolyfenceModule", code: 3, userInfo: [NSLocalizedDescriptionKey: "Invalid zone data"])
            }

            guard let tracker = locationTracker else {
                throw NSError(domain: "PolyfenceModule", code: 4, userInfo: [NSLocalizedDescriptionKey: "Location tracker not initialized"])
            }

            tracker.addZone(zoneId: zoneId, zoneName: zoneName, zoneData: zoneDict)
            sendStatus(trackingEnabled: nil)
            resolve(nil)
        } catch {
            NSLog("PolyfenceModule: Add zone failed: %@", error.localizedDescription)
            reject("ADD_ZONE_FAILED", error.localizedDescription, error)
        }
    }

    @objc(removeZone:resolver:rejecter:)
    func removeZone(zoneId: String?, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let zoneId = zoneId, !zoneId.isEmpty else {
                throw NSError(domain: "PolyfenceModule", code: 5, userInfo: [NSLocalizedDescriptionKey: "Zone ID is required"])
            }

            guard let tracker = locationTracker else {
                throw NSError(domain: "PolyfenceModule", code: 6, userInfo: [NSLocalizedDescriptionKey: "Location tracker not initialized"])
            }

            tracker.removeZone(zoneId: zoneId)
            sendStatus(trackingEnabled: nil)
            resolve(nil)
        } catch {
            NSLog("PolyfenceModule: Remove zone failed: %@", error.localizedDescription)
            reject("REMOVE_ZONE_FAILED", error.localizedDescription, error)
        }
    }

    @objc(removeAllZones:rejecter:)
    func removeAllZones(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let tracker = locationTracker else {
                throw NSError(domain: "PolyfenceModule", code: 7, userInfo: [NSLocalizedDescriptionKey: "Location tracker not initialized"])
            }

            tracker.clearAllZones()
            sendStatus(trackingEnabled: nil)
            resolve(nil)
        } catch {
            NSLog("PolyfenceModule: Clear zones failed: %@", error.localizedDescription)
            reject("CLEAR_ZONES_FAILED", error.localizedDescription, error)
        }
    }

    @objc(getZoneStates:rejecter:)
    func getZoneStates(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let tracker = locationTracker else {
                throw NSError(domain: "PolyfenceModule", code: 8, userInfo: [NSLocalizedDescriptionKey: "Location tracker not initialized"])
            }

            let states = tracker.getCurrentZoneStates()
            resolve(states)
        } catch {
            NSLog("PolyfenceModule: Get zone states failed: %@", error.localizedDescription)
            reject("ZONE_STATES_FAILED", error.localizedDescription, error)
        }
    }

    @objc(getDebugInfo:rejecter:)
    func getDebugInfo(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            let debugInfo = PolyfenceDebugCollector.shared.collectDebugInfo()
            resolve(debugInfo)
        } catch {
            NSLog("PolyfenceModule: Get debug info failed: %@", error.localizedDescription)
            reject("DEBUG_INFO_FAILED", error.localizedDescription, error)
        }
    }

    @objc(getSessionTelemetry:rejecter:)
    func getSessionTelemetry(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let tracker = locationTracker else {
                throw NSError(domain: "PolyfenceModule", code: 9, userInfo: [NSLocalizedDescriptionKey: "Location tracker not initialized"])
            }

            var telemetry = tracker.getSessionTelemetryData()
            telemetry["deviceCategory"] = Self.getDeviceCategory()
            telemetry["osVersionMajor"] = ProcessInfo.processInfo.operatingSystemVersion.majorVersion
            telemetry["chargingDuringSession"] = UIDevice.current.batteryState == .charging || UIDevice.current.batteryState == .full
            resolve(telemetry)
        } catch {
            NSLog("PolyfenceModule: Get session telemetry failed: %@", error.localizedDescription)
            reject("TELEMETRY_FAILED", error.localizedDescription, error)
        }
    }

    @objc(setTrackingSchedule:resolver:rejecter:)
    func setTrackingSchedule(schedule: NSDictionary?, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let scheduleDict = schedule as? [String: Any] else {
                throw NSError(domain: "PolyfenceModule", code: 10, userInfo: [NSLocalizedDescriptionKey: "Schedule is required"])
            }

            guard let tracker = locationTracker else {
                throw NSError(domain: "PolyfenceModule", code: 11, userInfo: [NSLocalizedDescriptionKey: "Location tracker not initialized"])
            }

            if let scheduleSettings = scheduleDict as? [String: Any] {
                tracker.setScheduleConfig(scheduleSettings)
            }

            resolve(nil)
        } catch {
            NSLog("PolyfenceModule: Set tracking schedule failed: %@", error.localizedDescription)
            reject("SCHEDULE_FAILED", error.localizedDescription, error)
        }
    }

    @objc(clearTrackingSchedule:rejecter:)
    func clearTrackingSchedule(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let tracker = locationTracker else {
                throw NSError(domain: "PolyfenceModule", code: 12, userInfo: [NSLocalizedDescriptionKey: "Location tracker not initialized"])
            }

            tracker.clearScheduleConfig()
            resolve(nil)
        } catch {
            NSLog("PolyfenceModule: Clear tracking schedule failed: %@", error.localizedDescription)
            reject("CLEAR_SCHEDULE_FAILED", error.localizedDescription, error)
        }
    }

    // MARK: - PolyfenceCoreDelegate Implementation

    func onLocationUpdate(_ locationData: [String: Any]) {
        sendLocationEvent(locationData)
    }

    func onGeofenceEvent(
        zoneId: String,
        zoneName: String,
        eventType: String,
        latitude: Double,
        longitude: Double,
        detectionTimeMs: Double,
        gpsAccuracy: Double,
        speedMps: Double,
        activityAtEvent: String,
        distanceToBoundaryM: Double
    ) {
        let event: [String: Any] = [
            "zoneId": zoneId,
            "zoneName": zoneName,
            "eventType": eventType,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
            "latitude": latitude,
            "longitude": longitude,
            "detectionTimeMs": detectionTimeMs,
            "gpsAccuracy": gpsAccuracy,
            "speedMps": speedMps,
            "activityAtEvent": activityAtEvent,
            "distanceToBoundaryM": distanceToBoundaryM
        ]
        sendGeofenceEvent(event)
    }

    func onError(_ errorCode: String, errorMessage: String, details: [String: Any]?) {
        var errorMap: [String: Any] = [
            "code": errorCode,
            "message": errorMessage,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000)
        ]
        if let details = details {
            errorMap.merge(details) { _, new in new }
        }
        sendErrorEvent(errorMap)
    }

    func onPerformanceEvent(_ event: [String: Any]) {
        sendPerformanceEvent(event)
    }

    // MARK: - Private Event Sending Methods

    private func sendLocationEvent(_ locationData: [String: Any]) {
        guard hasListeners else { return }
        sendEvent(withName: "onLocation", body: locationData)
    }

    private func sendGeofenceEvent(_ eventData: [String: Any]) {
        guard hasListeners else { return }
        sendEvent(withName: "onGeofenceEvent", body: eventData)
    }

    private func sendErrorEvent(_ errorData: [String: Any]) {
        guard hasListeners else { return }
        sendEvent(withName: "onError", body: errorData)
    }

    private func sendPerformanceEvent(_ eventData: [String: Any]) {
        guard hasListeners else { return }
        sendEvent(withName: "onPerformance", body: eventData)
    }

    private func sendStatus(trackingEnabled: Bool?) {
        let zonesCount = (try? zonePersistence?.getZoneCount()) ?? 0
        let payload: [String: Any?] = [
            "type": "status",
            "trackingEnabled": trackingEnabled ?? false,
            "zonesCount": zonesCount,
            "profile": nil,
            "lastAccuracy": nil,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000)
        ]
        sendPerformanceEvent(payload as [String: Any])
    }

    // MARK: - Device Category Detection

    private static func getDeviceCategory() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machine = withUnsafePointer(to: &systemInfo.machine) {
            $0.withMemoryRebound(to: CChar.self, capacity: 1) {
                String(validatingUTF8: $0) ?? "unknown"
            }
        }

        if machine.hasPrefix("iPhone") {
            let parts = machine.replacingOccurrences(of: "iPhone", with: "").split(separator: ",")
            if let first = parts.first {
                let modelNum = Int(first) ?? 0
                return categorizeIPhoneModel(modelNum)
            }
        } else if machine.hasPrefix("iPad") {
            return "ipad"
        }
        return "ios_other"
    }

    private static func categorizeIPhoneModel(_ modelNumber: Int) -> String {
        switch modelNumber {
        case 15, 16:
            return "iphone_15"
        case 14, 13:
            return "iphone_14_13"
        case 12, 11:
            return "iphone_12_11"
        default:
            return "iphone_older"
        }
    }
}

File: android/src/main/kotlin/io/polyfence/reactnative/PolyfenceModule.kt
package io.polyfence.reactnative

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.SharedPreferences
import android.location.LocationManager
import android.net.Uri
import android.os.Build
import android.os.PowerManager
import android.provider.Settings
import android.util.Log
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import io.polyfence.core.LocationTracker
import io.polyfence.core.PolyfenceCoreDelegate
import io.polyfence.core.PolyfenceErrorManager
import io.polyfence.core.PolyfenceDebugCollector
import io.polyfence.core.ZonePersistence
import io.polyfence.core.configuration.ActivitySettings
import io.polyfence.core.configuration.SmartGpsConfig
import io.polyfence.core.configuration.SmartGpsConfigFactory
import java.util.Locale

/**
 * React Native module for Polyfence geofencing bridge.
 * Implements PolyfenceCoreDelegate to receive events from LocationTracker.
 * Single responsibility: React ↔ LocationTracker communication.
 */
class PolyfenceModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext), PolyfenceCoreDelegate {

    companion object {
        private const val PREFS_NAME = "polyfence_state"
        private const val KEY_TRACKING_ENABLED = "tracking_enabled"
        private var pendingBridgePlatform: String? = null
    }

    private val context: Context = reactContext
    private var locationTracker: LocationTracker? = null

    override fun getName(): String = "Polyfence"

    @ReactMethod
    fun initialize(config: ReadableMap?, promise: Promise) {
        try {
            val configMap = config?.toHashMap() ?: mapOf()

            val version = (configMap["config"] as? Map<*, *>)?.get("pluginVersion") as? String
            if (version != null) {
                PolyfenceDebugCollector.setPluginVersion(version)
            }

            val disableAlerts = (configMap["config"] as? Map<*, *>)?.get("disableAlertNotifications") as? Boolean ?: false
            LocationTracker.setAlertNotificationsEnabled(!disableAlerts)

            locationTracker = LocationTracker.getInstance(context)
            locationTracker?.setDelegate(this)

            pendingBridgePlatform?.let { platform ->
                LocationTracker.setBridgePlatform(platform)
                pendingBridgePlatform = null
            } ?: run {
                LocationTracker.setBridgePlatform("react-native")
            }

            PolyfenceErrorManager.initialize { errorMap ->
                sendErrorEvent(errorMap)
            }

            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Initialization failed: ${e.message}")
            promise.reject("INITIALIZATION_FAILED", e.message)
        }
    }

    @ReactMethod
    fun startTracking(promise: Promise) {
        try {
            setTrackingEnabled(context, true)
            val intent = Intent(context, LocationTracker::class.java).apply {
                action = LocationTracker.ACTION_START_TRACKING
            }
            context.startForegroundService(intent)
            sendStatus(context)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to start tracking: ${e.message}")
            promise.reject("START_TRACKING_FAILED", e.message)
        }
    }

    @ReactMethod
    fun stopTracking(promise: Promise) {
        try {
            setTrackingEnabled(context, false)
            val intent = Intent(context, LocationTracker::class.java).apply {
                action = LocationTracker.ACTION_STOP_TRACKING
            }
            context.startService(intent)
            sendStatus(context)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to stop tracking: ${e.message}")
            promise.reject("STOP_TRACKING_FAILED", e.message)
        }
    }

    @ReactMethod
    fun addZone(zoneData: ReadableMap?, promise: Promise) {
        try {
            val zoneMap = zoneData?.toHashMap() ?: run {
                promise.reject("INVALID_ZONE", "Zone data is required")
                return
            }

            val zoneId = zoneMap["id"] as? String ?: run {
                promise.reject("INVALID_ZONE", "Zone ID is required")
                return
            }

            val zoneName = zoneMap["name"] as? String ?: "Unknown Zone"

            if (!isTrackingEnabled(context)) {
                try {
                    val persistence = ZonePersistence(context)
                    persistence.saveZone(zoneId, zoneName, zoneMap)
                } catch (e: Exception) {
                    Log.w("PolyfenceModule", "Failed to persist zone $zoneId: ${e.message}")
                }
                sendStatus(context)
                promise.resolve(null)
                return
            }

            val intent = Intent(context, LocationTracker::class.java).apply {
                action = LocationTracker.ACTION_ADD_ZONE
                putExtra("zoneId", zoneId)
                putExtra("zoneName", zoneName)
                putExtra("zoneData", HashMap(zoneMap))
            }
            context.startService(intent)
            sendStatus(context)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to add zone: ${e.message}")
            promise.reject("ADD_ZONE_FAILED", e.message)
        }
    }

    @ReactMethod
    fun removeZone(zoneId: String?, promise: Promise) {
        try {
            if (zoneId.isNullOrBlank()) {
                promise.reject("INVALID_ZONE_ID", "Zone ID is required")
                return
            }

            val intent = Intent(context, LocationTracker::class.java).apply {
                action = LocationTracker.ACTION_REMOVE_ZONE
                putExtra("zoneId", zoneId)
            }
            context.startService(intent)
            sendStatus(context)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to remove zone: ${e.message}")
            promise.reject("REMOVE_ZONE_FAILED", e.message)
        }
    }

    @ReactMethod
    fun removeAllZones(promise: Promise) {
        try {
            val intent = Intent(context, LocationTracker::class.java).apply {
                action = LocationTracker.ACTION_CLEAR_ZONES
            }
            context.startService(intent)
            sendStatus(context)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to clear all zones: ${e.message}")
            promise.reject("CLEAR_ZONES_FAILED", e.message)
        }
    }

    @ReactMethod
    fun getZoneStates(promise: Promise) {
        try {
            val states = LocationTracker.getCurrentZoneStates()
            promise.resolve(states)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to get zone states: ${e.message}")
            promise.reject("ZONE_STATES_FAILED", e.message)
        }
    }

    @ReactMethod
    fun getDebugInfo(promise: Promise) {
        try {
            val debugInfo = PolyfenceDebugCollector.collectDebugInfo(context)
            promise.resolve(debugInfo)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to get debug info: ${e.message}")
            promise.reject("DEBUG_INFO_FAILED", e.message)
        }
    }

    @ReactMethod
    fun getSessionTelemetry(promise: Promise) {
        try {
            val telemetry = LocationTracker.getSessionTelemetry()
            val sessionData = HashMap(telemetry)
            sessionData["deviceCategory"] = getDeviceCategory()
            sessionData["osVersionMajor"] = Build.VERSION.SDK_INT
            promise.resolve(sessionData)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to get session telemetry: ${e.message}")
            promise.reject("TELEMETRY_FAILED", e.message)
        }
    }

    @ReactMethod
    fun setTrackingSchedule(schedule: ReadableMap?, promise: Promise) {
        try {
            if (schedule == null) {
                promise.reject("INVALID_SCHEDULE", "Schedule is required")
                return
            }

            val scheduleMap = schedule.toHashMap()
            val intent = Intent(context, LocationTracker::class.java).apply {
                action = LocationTracker.ACTION_UPDATE_CONFIG
                putExtra("schedule", HashMap(scheduleMap))
            }
            context.startService(intent)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to set tracking schedule: ${e.message}")
            promise.reject("SCHEDULE_FAILED", e.message)
        }
    }

    @ReactMethod
    fun clearTrackingSchedule(promise: Promise) {
        try {
            val intent = Intent(context, LocationTracker::class.java).apply {
                action = LocationTracker.ACTION_UPDATE_CONFIG
                putExtra("clearSchedule", true)
            }
            context.startService(intent)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to clear tracking schedule: ${e.message}")
            promise.reject("CLEAR_SCHEDULE_FAILED", e.message)
        }
    }

    /**
     * PolyfenceCoreDelegate implementation: receive location updates from LocationTracker
     */
    override fun onLocationUpdate(locationData: Map<String, Any>) {
        sendLocationEvent(locationData)
    }

    /**
     * PolyfenceCoreDelegate implementation: receive geofence events from LocationTracker
     */
    override fun onGeofenceEvent(
        zoneId: String,
        zoneName: String,
        eventType: String,
        latitude: Double,
        longitude: Double,
        detectionTimeMs: Double,
        gpsAccuracy: Double,
        speedMps: Double,
        activityAtEvent: String,
        distanceToBoundaryM: Double
    ) {
        val event = mapOf(
            "zoneId" to zoneId,
            "zoneName" to zoneName,
            "eventType" to eventType,
            "timestamp" to System.currentTimeMillis(),
            "latitude" to latitude,
            "longitude" to longitude,
            "detectionTimeMs" to detectionTimeMs,
            "gpsAccuracy" to gpsAccuracy,
            "speedMps" to speedMps,
            "activityAtEvent" to activityAtEvent,
            "distanceToBoundaryM" to distanceToBoundaryM
        )
        sendGeofenceEvent(event)
    }

    /**
     * PolyfenceCoreDelegate implementation: receive error events from LocationTracker
     */
    override fun onError(errorCode: String, errorMessage: String, details: Map<String, Any>?) {
        val errorMap = mutableMapOf(
            "code" to errorCode,
            "message" to errorMessage,
            "timestamp" to System.currentTimeMillis()
        )
        if (details != null) {
            errorMap.putAll(details)
        }
        sendErrorEvent(errorMap)
    }

    /**
     * PolyfenceCoreDelegate implementation: receive performance events from LocationTracker
     */
    override fun onPerformanceEvent(event: Map<String, Any>) {
        sendPerformanceEvent(event)
    }

    /**
     * Send location event via onLocation event name
     */
    private fun sendLocationEvent(locationData: Map<String, Any>) {
        try {
            val event = Arguments.createMap()
            for ((key, value) in locationData) {
                when (value) {
                    is String -> event.putString(key, value)
                    is Int -> event.putInt(key, value)
                    is Double -> event.putDouble(key, value)
                    is Boolean -> event.putBoolean(key, value)
                    is Long -> event.putDouble(key, value.toDouble())
                    null -> event.putNull(key)
                    else -> event.putString(key, value.toString())
                }
            }
            sendEvent("onLocation", event)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to send location event: ${e.message}")
        }
    }

    /**
     * Send geofence event via onGeofenceEvent event name
     */
    private fun sendGeofenceEvent(eventData: Map<String, Any>) {
        try {
            val event = Arguments.createMap()
            for ((key, value) in eventData) {
                when (value) {
                    is String -> event.putString(key, value)
                    is Int -> event.putInt(key, value)
                    is Double -> event.putDouble(key, value)
                    is Boolean -> event.putBoolean(key, value)
                    is Long -> event.putDouble(key, value.toDouble())
                    null -> event.putNull(key)
                    else -> event.putString(key, value.toString())
                }
            }
            sendEvent("onGeofenceEvent", event)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to send geofence event: ${e.message}")
        }
    }

    /**
     * Send error event via onError event name
     */
    private fun sendErrorEvent(errorData: Map<String, Any>) {
        try {
            val event = Arguments.createMap()
            for ((key, value) in errorData) {
                when (value) {
                    is String -> event.putString(key, value)
                    is Int -> event.putInt(key, value)
                    is Double -> event.putDouble(key, value)
                    is Boolean -> event.putBoolean(key, value)
                    is Long -> event.putDouble(key, value.toDouble())
                    null -> event.putNull(key)
                    else -> event.putString(key, value.toString())
                }
            }
            sendEvent("onError", event)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to send error event: ${e.message}")
        }
    }

    /**
     * Send performance event via onPerformance event name
     */
    private fun sendPerformanceEvent(eventData: Map<String, Any>) {
        try {
            val event = Arguments.createMap()
            for ((key, value) in eventData) {
                when (value) {
                    is String -> event.putString(key, value)
                    is Int -> event.putInt(key, value)
                    is Double -> event.putDouble(key, value)
                    is Boolean -> event.putBoolean(key, value)
                    is Long -> event.putDouble(key, value.toDouble())
                    null -> event.putNull(key)
                    else -> event.putString(key, value.toString())
                }
            }
            sendEvent("onPerformance", event)
        } catch (e: Exception) {
            Log.e("PolyfenceModule", "Failed to send performance event: ${e.message}")
        }
    }

    /**
     * Helper: send event to React Native listeners
     */
    private fun sendEvent(eventName: String, params: WritableMap) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    /**
     * Send status event (zone count, tracking enabled) to onPerformance
     */
    private fun sendStatus(context: Context) {
        val tracking = isTrackingEnabled(context)
        val zonesCount = try {
            val persistence = ZonePersistence(context)
            persistence.getZoneCount()
        } catch (e: Exception) { 0 }

        val statusMap = Arguments.createMap().apply {
            putString("type", "status")
            putBoolean("trackingEnabled", tracking)
            putInt("zonesCount", zonesCount)
            putNull("profile")
            putNull("lastAccuracy")
            putDouble("timestamp", System.currentTimeMillis().toDouble())
        }
        sendEvent("onPerformance", statusMap)
    }

    /**
     * Device category bucketing (not exact model) for ML telemetry
     */
    private fun getDeviceCategory(): String {
        val manufacturer = Build.MANUFACTURER.lowercase(Locale.ROOT)
        val model = Build.MODEL.lowercase(Locale.ROOT)
        return when {
            manufacturer.contains("samsung") -> when {
                model.contains("sm-s9") || model.contains("sm-s24") || model.contains("sm-s23") || model.contains("sm-f") -> "samsung_flagship"
                model.contains("sm-a5") || model.contains("sm-a7") || model.contains("sm-a3") -> "samsung_mid"
                else -> "samsung_other"
            }
            manufacturer.contains("google") || manufacturer.contains("pixel") -> "google_pixel"
            manufacturer.contains("xiaomi") || manufacturer.contains("redmi") -> "xiaomi"
            manufacturer.contains("huawei") -> "huawei"
            manufacturer.contains("oneplus") -> "oneplus"
            manufacturer.contains("oppo") -> "oppo"
            manufacturer.contains("vivo") -> "vivo"
            else -> "android_other"
        }
    }

    /**
     * Tracking state helpers
     */
    private fun isTrackingEnabled(context: Context): Boolean {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        return prefs.getBoolean(KEY_TRACKING_ENABLED, false)
    }

    private fun setTrackingEnabled(context: Context, enabled: Boolean) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit().putBoolean(KEY_TRACKING_ENABLED, enabled).apply()
    }
}
